### ISSUE TITLE
Variational Model

### ISSUE THREAD

**Issue by MatteoRobbiati** [qiboteam/qibo#757]
I think it should be great to have a general object to implement variational models. 
It can be something inheriting the circuit features and override some methods in order to simplify some training procedures. 
This can be useful, for example, when we want to calculate gradients of a loss function with respect  the variational parameters which are combined together with an external input data (think about re-uploading models).

Doing this, we might think about the possibility of "tensoring" actions involving parameters (`set_parameters`, `upload_x`, etc.).

**Comment by scarrazza** [comment]
@MatteoRobbiati shall we close this?
